### ISSUE TITLE
netdata does not send notifications for alarms which fail with (errno 12, Out of memory)

### ISSUE THREAD

**Issue by svrnwnsch** [netdata/netdata#6335]
##### Bug report summary
I am running Netdata in a docker container. After several days of uninterrupted running I have noticed that at some time notifications are no longer send to my services. In the docker logs I see the following errors (the errors where cause by restarting another server which streams to this server):
```
netdata_1  | 2019-06-24 09:18:38: netdata ERROR : HEALTH : HEALTH: Cannot popen("exec /usr/libexec/netdata/plugins.d/alarm-notify.sh 'silent' 'XXXXXXXX' '1560932661' '1560931251' '8' '1561367918' 'lowest_entropy' 'system.entropy' 'entropy' 'WARNING' 'CLEAR' '35' '3228' '6@/usr/lib/netdata/conf.d/health.d/entropy.conf' '355200' '0' 'entries' 'minimum entries in the random numbers pool in the last 10 minutes' '35 entries' '3228 entries' '$this < (($status >= $WARNING) ? (200) : (100))' '[ $this = 35 ] [ $status = 1 ] [ $WARNING = 3 ] ' '2' '0'", "r"). (errno 12, Out of memory)
....
netdata_1  | 2019-06-24 09:18:40: STREAM: 2229 '[10.0.1.4]:51036' 'DISCONNECTED' host 'XXXXXXXX' api key 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXX' machine guid 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
netdata_1  | 2019-06-24 09:18:40: netdata ERROR : STREAM_RECEIVER[XXXXXXXX,[10.0.1.4]:51036] : read failed
netdata_1  | 2019-06-24 09:18:40: netdata ERROR : STREAM_RECEIVER[XXXXXXXX,[10.0.1.4]:51036] : STREAM XXXXXXXX [receive from [10.0.1.4]:51036]: disconnected (completed 1965197 updates). (errno 22, Invalid argument)
netdata_1  | 2019-06-24 09:18:40: netdata INFO  : STREAM_RECEIVER[XXXXXXXX,[10.0.1.4]:51036] : STREAM XXXXXXXX [receive from [10.0.1.4]:51036]: receive thread ended (task id 2229)
netdata_1  | 2019-06-24 09:18:40: netdata INFO  : STREAM_RECEIVER[XXXXXXXX,[10.0.1.4]:51036] : thread with task id 2229 finished
....
netdata_1  | 2019-06-24 09:18:53: netdata ERROR : HEALTH : HEALTH: Cannot popen("exec /usr/libexec/netdata/plugins.d/alarm-notify.sh 'sysadmin' 'XXXXXXXX' '1560932664' '1560931246' '6' '1561367933' 'ipv4_tcphandshake_last_collected_secs' 'ipv4.tcphandshake' 'tcp' 'WARNING' 'CLEAR' '15' '5' '6@/usr/lib/netdata/conf.d/health.d/tcp_resets.conf' '436215' '0' 'seconds ago' 'number of seconds since the last successful data collection' '00:00:15 ago' '00:00:05 ago' '$this > (($status >= $WARNING) ? ($update_every) : ( 5 * $update_every))' '[ $this = 15 ] [ $status = 1 ] [ $WARNING = 3 ] [ ${update_every} = 2 ] ' '5' '0'", "r"). (errno 12, Out of memory)
netdata_1  | 2019-06-24 09:18:53: netdata ERROR : HEALTH : HEALTH: Cannot popen("exec /usr/libexec/netdata/plugins.d/alarm-notify.sh 'sysadmin' 'XXXXXXXX' '1560932663' '1560931236' '6' '1561367933' 'ipv4_udperrors_last_collected_secs' 'ipv4.udperrors' 'udp' 'WARNING' 'CLEAR' '15' '5' '6@/usr/lib/netdata/conf.d/health.d/udp_errors.conf' '436215' '0' 'seconds ago' 'number of seconds since the last successful data collection' '00:00:15 ago' '00:00:05 ago' '$this > (($status >= $WARNING) ? ($update_every) : ( 5 * $update_every))' '[ $this = 15 ] [ $status = 1 ] [ $WARNING = 3 ] [ ${update_every} = 2 ] ' '5' '0'", "r"). (errno 12, Out of memory)
netdata_1  | 2019-06-24 09:18:53: netdata ERROR : HEALTH : HEALTH: Cannot popen("exec /usr/libexec/netdata/plugins.d/alarm-notify.sh 'sysadmin' 'XXXXXXXX' '1560932662' '1560931229' '6' '1561367933' 'netfilter_last_collected_secs' 'netfilter.conntrack_sockets' 'conntrack' 'WARNING' 'CLEAR' '15' '5' '4@/usr/lib/netdata/conf.d/health.d/netfilter.conf' '436215' '0' 'seconds ago' 'number of seconds since the last successful data collection' '00:00:15 ago' '00:00:05 ago' '$this > (($status >= $WARNING) ? ($update_every) : ( 5 * $update_every))' '[ $this = 15 ] [ $status = 1 ] [ $WARNING = 3 ] [ ${update_every} = 2 ] ' '5' '0'", "r"). (errno 12, Out of memory)
```
In the logs netdata states it is out of memory. In contrast to this netdata shows me that the server it is running on only uses 81% of 4gb of memory, and there are ~500 MiB available to the system (mem.available).

Restarting the docker container seems to solve the problem. Nevertheless it highly annoying that the notifications fail here. Additionally if you brows to the node which causes the unsent alarms, the notifications are displayed in the browser.

##### OS / Environment
Ubuntu, running netdata in docker-container pulled from https://hub.docker.com/r/netdata/netdata/

##### Netdata version (ouput of `netdata -V`)
v1.15.0-103-gc5ab8255

##### Component Name
idk

##### Steps To Reproduce
Run netdata for multiple days within docker and stream data to it.
Trigger an alarm.

##### Expected behavior
Do not fail sending error messages when there is still memory left. Or free memory elsewhere as alarm messages are very important to use netdata for monitoring servers.

**Comment by svrnwnsch** [comment]
As I try to understand what popen does it looks to me that it forks the process which I would guess would then result in the doubled used memory?
If this is the case it is clear to me why this fails as netdata is using on the server currently 1,7gb as reported by: apps.mem

**Comment by cakrit** [comment]
Well, the out of memory error can't be faked. Can you please send me a snapshot from the netdata UI of the server where you saw it, so I can see that mem.available chart and others, like the netdata memory usage from the application charts?  (email it to chris@netdata.cloud). 

I really like this idea: 
> free memory elsewhere as alarm messages are very important to use netdata for monitoring servers.

Not easy of course, but it totally makes sense to put netdata in a "low memory mode", when the system is under pressure. We'll need to discuss it of course...

I didn't know that about `popen. I read https://stackoverflow.com/questions/46574798/enomem-from-popen-for-system-while-there-is-enough-memory and it's not good. 

Let us have an internal discussion and we'll come back. Do send me the snapshot if you can get it...

**Comment by cakrit** [comment]
Could you please tell me what "memory mode" you're running with?

**Comment by cakrit** [comment]
Another question, do you have a memory limit on the docker container? Because the mem.available you see is for the host running the container. You'd need to look at the cgroup and the limits there. I can tell for sure, if I get the snapshot.

**Comment by svrnwnsch** [comment]
> Could you please tell me what "memory mode" you're running with?

the memory mode running is dbengine.

> Another question, do you have a memory limit on the docker container? Because the mem.available you see is for the host running the container. You'd need to look at the cgroup and the limits there. I can tell for sure, if I get the snapshot.

I run netdata via docker-compose with now self set limitations. If I run `docker stats` the following is presented:
```
CONTAINER ID        NAME                CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
7887059f6c97        user_caddy_1     0.01%               16.93MiB / 3.828GiB   0.43%               201MB / 227MB       24.7MB / 4.1kB      23
e35ba54f6f18        user_netdata_1   3.57%               2.54GiB / 3.828GiB    66.36%              236GB / 11.2GB      5.28GB / 3.43GB     52

```
@cakrit Can I somehow send you the snapshot private, to not disclose any private information?

**Comment by mfundul** [comment]
Memory mode `dbengine` will increase the initial memory footprint so it adds insult to injury with `popen()`.

**Comment by cakrit** [comment]
> @cakrit Can I somehow send you the snapshot private, to not disclose any private information?

That's why I asked you to email it to me. :)
chris@netdata.cloud

**Comment by mfundul** [comment]
@svrnwnsch can you test the following workaround?
[Here](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/7/html/performance_tuning_guide/sect-red_hat_enterprise_linux-performance_tuning_guide-configuration_tools-configuring_system_memory_capacity) and [here](https://www.kernel.org/doc/Documentation/sysctl/vm.txt) is information regarding `overcommit_memory`.

Can you verify that if you run `echo 1 > /proc/sys/vm/overcommit_memory` as root netdata works and the rest of the system does not run out of memory? In theory, `fork()` will duplicate netdata memory but only before it is overwritten by the `exec()` family of functions, which should get rid of the any duplicate memory thanks to the kernel's Copy-On-Write mechanisms.

This means that netdata exceeds the memory limits only for a fragment of a second, and in a virtual way and not by actually allocating resident memory. This will help point us to the right direction. Thank you.

**Comment by svrnwnsch** [comment]
> Can you verify that if you run echo 1 > /proc/sys/vm/overcommit_memory as root netdata works and the rest of the system does not run out of memory?

Should I run this within the netdata container or directly on the server?

**Comment by mfundul** [comment]
> 
> 
> > Can you verify that if you run echo 1 > /proc/sys/vm/overcommit_memory as root netdata works and the rest of the system does not run out of memory?
> 
> Should I run this within the netdata container or directly on the server?

You should run it directly on the system since not all kernel functionality is containerized.

**Comment by svrnwnsch** [comment]
Changing this value fixes the problem that no messages are send.
And the server does not run out of memory (at leas not in the last 1h since I change it).

**Comment by cakrit** [comment]
Cool and thanks for the export. These two things verify our root cause analysis and @mfundul has already started on a fix to address the issue properly (#6339).
